teps/0044-composing-tasks-with-tasks.md

```diff
@@ -0,0 +1,135 @@
+Tasks in a Pipeline and they need to share data (beyond a simple
+[result](https://github.com/tektoncd/pipeline/blob/master/docs/pipelines.md#using-results)),
+you'll need to provision a PVC or do some other similar, cloud specific thing,
+to [make a volume available](https://github.com/tektoncd/pipeline/blob/master/docs/workspaces.md#specifying-volumesources-in-workspaces)
+that can be shared between them.
+
+Some problems with this:
+* PVCs add additional overhead, both in speed and in management (somewhat
+  mitigated by
+  [letting tekton create and delete volumes for you](https://github.com/tektoncd/pipeline/blob/master/docs/workspaces.md#volumeclaimtemplate)
+* Sometimes you just want to run a Task with a teeny bit more functionality,
+  e.g. you want to use a task to build images, but you want to clone source
+  with git first, and it seems like overkill to have to construct a pipeline
+
+### PipelineResources
+
+This kind of compositional functionality was being (somewhat! read on!) provided
+by [PipelineResources](https://github.com/tektoncd/pipeline/blob/master/docs/resources.md).
+[We did not bring these types to beta](https://github.com/tektoncd/pipeline/blob/master/docs/resources.md#why-arent-pipelineresources-in-beta).
+Issues that don't let PipelineResources (as is) solve these problems are:
+* PipelineResource "outputs" do not run when the steps of the Task fail (or if
+  a previous PipelienResource "output" fails) (see [unit test use case](#use-cases-optional))
+* PipelineResources do not use Tasks (which is something we could change), meaning
+  you cannot use them to compose Tasks together, you need to build PipelineResources
+  * Adding new PipelineResources currently needs to be done in the Tekton Pipelines controller,
+    tho there have been several attempts to propose ways to make it extensible:
+    * [PipelineResources 2 Uber Design Doc](https://docs.google.com/document/d/1euQ_gDTe_dQcVeX4oypODGIQCAkUaMYQH5h7SaeFs44/edit)
+    * [Tekton Pipeline Resource Extensibility](https://docs.google.com/document/d/1rcMG1cIFhhixMSmrT734MBxvH3Ghre9-4S2wbzodPiU/edit)
+  * All of this is even though PipelineResources are _very_ similar to Tasks in that they are
+    collections of steps (which [get injected before and after a Task's steps](https://github.com/tektoncd/pipeline/issues/1838)
+* Tasks have to declare in advance what PipelineResources they need; you can't decide to use
+  a PipelineResource with a Task after it has been written (e.g. in a TaskRun or a Pipeline)
+  and you can't mix types of PipelineResource (e.g. if a Task declares it needs a
+  [git PipelineResources](https://github.com/tektoncd/pipeline/blob/master/docs/resources.md#git-resource)
+  it can't use a different kind of PipelineResource to provide those files (tho you can avoid
+  doing the clone multiple times using [from](https://github.com/tektoncd/pipeline/blob/master/docs/pipelines.md#using-the-from-parameter))
+
```

Comment:
Finding a path forward for the PipelineResources use-cases is really important to us. If a mutating admission webhook could be used to transform existing use of PipelineResources into a pre and post Task that would provide a backwards compatible story while letting the runtime remove support.

Reply:
Hmm... I wonder if the Conditions case where it needs to use an image could also be re-cast as a pre-Task using a mutating admission webhook to let us also remove Conditions from the runtime.

Reply:
I've added a point under Goals about providing a migration path.

I'd caution though that pre and post Tasks is only one possible solution, another possible solution is to provide a scheduling mechanism that lets Tasks run together on the same pod (e.g. a Pipeline on one pod).

But I'm getting ahead of myself - we can start exploring solutions once we get this problem statement merged :D

Reply:
Yes, this shouldn't be a primary goal for this TEP. It is important that we provide a migration part, but we don't have to overly think about it now.